### PR TITLE
Windows Terminal Support

### PR DESCRIPTION
**Pull request by daxgames**
# Windows Terminal Support

First looks at Windows Terminal Integration

image

## NEED TO FIX

  - Command line help is for ConEmu only and needs to be updated.
  - Probably more.

## NEED TO ADD:

  - Nothing that I know of.
  - Maybe more.

### `Cmder.exe` Launcher:
  - Runs `%cmder_root%\vendor\windows-terminal\WindowsTerminal.exe` instead of ConEmu if `%cmder_root%\vendor\windows-terminal` exists.
  - Supports all existing command line arguments but `/title`, and `/single`.
  - Pass ALL arguments following `--` argument directly to `WindowsTerminal.exe` to use the full power of Windows Terminal from the command line.
  - Modified Cmder.bat to launch windows terminal by default if it exists and put the default config file where it needs to be before launching Windows Terminal

### Scripts:
  - Added an argument to build.ps1 that allows building with Windows terminal instead of Conemu.

### Config:
  - Created a '%cmder_root%\vendor\WindowsTerminal_settings_default.json' that results in the WindowsTerminal Session and profiles shown above. NOTE: Non-Cmder sessions are auto-generated by Windows Terminal.
    - All sessions work as they do today in Cmder using Conemu.
  - Changes to the word separator settings so it works more m=like we would expect it ti work.
  - Added Windows Terminal to `%cmder_root%\vendor\sources.json`

### CODE REVIEW

**Comment by daxgames** [comment]
We have downloadable Cmder with WIndows Terminal artifacts [here](https://github.com/cmderdev/cmder/actions/runs/6304390270?pr=2877)

**Comment by daxgames** [comment]
This is a first stab at making a Windows Terminal integrated Cmder distribution.

It builds both Conemu and Windows Terminal Cmder archives.

I am 100% positive the launcher code could be better, but I am not a C++ developer.  It works and that is an accomplishment in itself.

File: vendor/windows_terminal_default_settings.json [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+                "startingDirectory": "",
+                "tabTitle": "Cmder as Admin"
+            },
+            {
+                "colorScheme": "Campbell",
+                "commandline": "PowerShell -ExecutionPolicy Bypass -NoLogo -NoProfile -NoExit -Command \"Invoke-Expression 'Import-Module ''%CMDER_ROOT%\\vendor.\\profile.ps1'''\"",
+                "guid": "{eb1f6578-ce9d-47a9-a8c7-9b3fdd22302d}",
+                "hidden": false,
+                "icon": "%CMDER_ROOT%\\icons\\cmder_orange.ico",
+                "name": "Cmder - Powershell",
```
Comment (Stanzilla) [inline vendor/windows_terminal_default_settings.json:80]:
```suggestion
                "name": "Cmder - PowerShell",
```

File: vendor/windows_terminal_default_settings.json [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+                "useAtlasEngine": false
+            },
+            {
+                "colorScheme": "Campbell",
+                "commandline": "PowerShell -ExecutionPolicy Bypass -NoLogo -NoProfile -NoExit -Command \"Invoke-Expression 'Import-Module ''%CMDER_ROOT%\\vendor.\\profile.ps1'''\"",
+                "elevate": true,
+                "guid": "{c5225c3e-8619-4145-8182-2800814eeb17}",
+                "hidden": false,
+                "icon": "%CMDER_ROOT%\\icons\\cmder_purple.ico",
+                "name": "Cmder - Powershell as Admin",
```
Comment (Stanzilla) [inline vendor/windows_terminal_default_settings.json:92]:
```suggestion
                "name": "Cmder - PowerShell as Admin",
```

Comment (Stanzilla) [inline vendor/windows_terminal_default_settings.json:92]:
```suggestion
                "name": "Cmder - PowerShell as Admin",
```

File: launcher/src/CmderLauncher.cpp [hunk 3]
```diff
@@ -465,8 +565,8 @@ void StartCmder(std::wstring  path = L"", bool is_single_mode = false, std::wstr
 	si.dwFlags = STARTF_TITLEISAPPID;
 #endif
 	PROCESS_INFORMATION pi;
-	if (!CreateProcess(conEmuPath, args, NULL, NULL, false, 0, NULL, NULL, &si, &pi)) {
-		MessageBox(NULL, _T("Unable to create the ConEmu process!"), _T("Error"), MB_OK);
+	if (!CreateProcess(terminalPath, args, NULL, NULL, false, 0, NULL, NULL, &si, &pi)) {
+		MessageBox(NULL, _T("Unable to create the terminal process!"), _T("Error"), MB_OK);
```
Comment (Stanzilla) [inline launcher/src/CmderLauncher.cpp:569]: might want to include the name of the terminal process so you know which failed?

File: Cmder.bat [hunk 4]
```diff
@@ -1,20 +1,66 @@
...
+  if exist "%~1" (
+      start %cmder_root%\vendor\conemu-maximus5\ConEmu.exe /Icon "%CMDER_ROOT%\icons\cmder.ico" /Title Cmder /LoadCfgFile "%~1"
+  ) else (
+      start %cmder_root%\vendor\conemu-maximus5\ConEmu.exe /Icon "%CMDER_ROOT%\icons\cmder.ico" /Title Cmder /LoadCfgFile "%CMDER_ROOT%\config\user_ConEmu.xml"
+  )
+  exit /b
+
+:windows-terminal
+  if not exist "%CMDER_ROOT%\vendor\windows-terminal\settings" md "%CMDER_ROOT%\vendor\windows-terminal\settings" 2>nul
+  if not exist "%CMDER_ROOT%\vendor\windows-terminal\.portable" echo "This make this installation of Windows Terminal portable" >"%CMDER_ROOT%\vendor\windows-terminal\.portable" 2>nul
```
Comment (Stanzilla) [inline Cmder.bat:41]:
```suggestion
  if not exist "%CMDER_ROOT%\vendor\windows-terminal\.portable" echo "This makes this installation of Windows Terminal portable" >"%CMDER_ROOT%\vendor\windows-terminal\.portable" 2>nul
```

**Review by Stanzilla — CHANGES_REQUESTED** [review]
pretty cool, haven't tested the code yet though and only have boring comments

**Comment by daxgames** [comment]
I have been using Cmder in Windows Terminal exclusively for about a month and I just decided it was time to start this.

It was actually easier than I thought it would be.

**Comment by daxgames** [comment]
I identified an error with Windows terminal support and profiles set to run as Administrator.  The `%CMDER_ROOT%` env var is not set and we have no way of figuring out what the root path for Cmder is on Administrative profiles.

I have requested a feature in Windows Terminal https://github.com/microsoft/terminal/issues/16295 that has the potential to resolve this the same way ConEmu does.  Our init scripts could then look for `%WT_FOLDER_PATH%` the same way we do for `%ConEmuDir%`

**Comment by MartiUK** [comment]
> I identified an error with Windows terminal support and profiles set to run as Administrator. The `%CMDER_ROOT%` env var is not set and we have no way of figuring out what the root path for Cmder is on Administrative profiles.
> 
> I have requested a feature in Windows Terminal [microsoft/terminal#16295](https://github.com/microsoft/terminal/issues/16295) that has the potential to resolve this the same way ConEmu does. Our init scripts could then look for `%WT_FOLDER_PATH%` the same way we do for `%ConEmuDir%`

Looks like it's slated for the WT 1.20 release, are you happy to wait or shall we try and find another solution?

**Comment by daxgames** [comment]
We can wait

**Comment by daxgames** [comment]
For those that want admin profiles before then you can set `CMDER_ROOT` globally via Windows env vars to fix the issue.

File: Cmder.bat [hunk 5]
```diff
@@ -1,20 +1,66 @@
...
-@if "%CMDER_ROOT:~-1%" == "\" SET CMDER_ROOT=%CMDER_ROOT:~0,-1%
-
-if not exist "%CMDER_ROOT%\config\user_ConEmu.xml" (
-    if not exist "%CMDER_ROOT%\config" mkdir "%CMDER_ROOT%\config" 2>nul
-    copy "%CMDER_ROOT%\vendor\ConEmu.xml.default" "%CMDER_ROOT%\config\user_ConEmu.xml" 1>nul
-    if %errorlevel% neq 0 (
-        echo ERROR: CMDER Initialization has Failed
-        exit /b 1
-    )
+set CMDER_TERMINAl=conemu
```
Comment (DRSDavidSoft) [inline Cmder.bat:5]: Please change the `CMDER_TERMINAl` to `CMDER_TERMINAL` (notice the lower case `l` at the end)

File: scripts/build.ps1 [hunk 6]
```diff
@@ -55,13 +55,16 @@ Param(
 
     # Using this option will skip all downloads, if you only need to build launcher
     [switch]$noVendor,
+    
+    # Using this option will specify the emulator to use [all, conemu-maximus5, or windows-terminal]
+    [string]$emulator = 'all',
```
Comment (DRSDavidSoft) [inline scripts/build.ps1:60]: @daxgames Would `$terminal` be a more claritive name to use for this switch?

File: scripts/build.ps1 [hunk 7]
```diff
@@ -55,13 +55,16 @@ Param(
...
 
     # Build launcher if you have MSBuild tools installed
     [switch]$Compile
 )
 
 # Get the scripts and cmder root dirs we are building in.
-$cmder_root = Resolve-Path "$PSScriptRoot\.."
+$cmder_root = [string](Resolve-Path "$PSScriptRoot\..")
```
Comment (DRSDavidSoft) [inline scripts/build.ps1:67]: I approve but just curious, do you remember what issue did you run into that required a cast to string?

File: scripts/build.ps1 [hunk 8]
```diff
@@ -132,6 +153,16 @@ if (-not $noVendor) {
         Download-File -Url $s.url -File $vend\$tempArchive -ErrorAction Stop
         Extract-Archive $tempArchive $s.name
 
+        # Make Embedded Windows Terminal Portable
+        if ($s.name -eq "windows-terminal") {
+          $windowTerminalFiles = resolve-path ($saveTo + "\" + $s.name + "\terminal*")
+					move-item -ErrorAction SilentlyContinue $windowTerminalFiles\* $s.name >$null
+					remove-item -ErrorAction SilentlyContinue $windowTerminalFiles >$null
```
Comment (DRSDavidSoft) [inline scripts/build.ps1:160]: Can you please correct the indentation for these two lines? (Tab/Space mismatch) Sorry for the nitpicking

File: scripts/pack.ps1 [hunk 9]
```diff
@@ -31,6 +31,9 @@ Param(
     # Path to the vendor configuration source file
     [string]$cmderRoot = "$PSScriptRoot\..",
 
+    # Using this option will pack artifacts for a specific included terminal emulator [all, conemu-maximus5, or windows-terminal]
+    [string]$emulator = 'all',
```
Comment (DRSDavidSoft) [inline scripts/pack.ps1:35]: ditto

File: vendor/init.bat [hunk 10]
```diff
@@ -299,17 +315,22 @@ if exist "%CMDER_ROOT%\vendor\git-for-windows" (
 )
 
 :SPECIFIED_GIT
-%print_debug% init.bat "Using /GIT_INSTALL_ROOT..."
+%print_debug% init.bat "Using specified GIT_INSTALL_ROOT=%GIT_INSTALL_ROOT%...."
 goto :CONFIGURE_GIT
 
 :FOUND_GIT
-%print_debug% init.bat "Using found Git '%GIT_VERSION_USER%' from '%GIT_INSTALL_ROOT%..."
+%print_debug% init.bat "Using found Git '%GIT_VERSION_USER%' from 'v%GIT_INSTALL_ROOT%..."
```
Comment (DRSDavidSoft) [inline vendor/init.bat:322]: @daxgames Can you please explain why `v` was prefixed? I'm confused

File: vendor/init.bat [hunk 11]
```diff
@@ -362,23 +397,41 @@ if defined git_locale (
...
-if defined CMDER_USER_BIN (
+%print_debug% init.bat "START - bin(prepend): Env Var - PATH=%path%"
+if %max_depth% gtr 1 (
+  %lib_path% enhance_path_recursive "%CMDER_ROOT%\bin" 0 %max_depth%
+) else (
+  set "path=%CMDER_ROOT%\bin;%path%"
+)
+%print_debug% init.bat "END - bin(prepend): Env Var - PATH=%path%"
+
+
```
Comment (DRSDavidSoft) [inline vendor/init.bat:421]: Is the double new line intentional ?

**Comment by DRSDavidSoft** [comment]
@daxgames I've merged this and the speed optimization PRs into development branch for testing purposes.
So sorry it's taking this long to review, I'm still unimaginably busy with some projects and IRL issues, but I'm super eager to test these great additions and apply them in master soon.
Hoping for a speedy Cmder 1.4 release!

File: vendor/init.bat (hunk 10 above)
Comment (daxgames) [reply vendor/init.bat:322]: No idea.

File: scripts/build.ps1 (hunk 6 above)
Comment (daxgames) [reply scripts/build.ps1:60]: Changed to Terminal

File: vendor/init.bat (hunk 10 above)
Comment (daxgames) [reply vendor/init.bat:322]: Fixed

File: vendor/init.bat (hunk 11 above)
Comment (daxgames) [reply vendor/init.bat:421]: No - fixed

File: Cmder.bat (hunk 5 above)
Comment (daxgames) [reply Cmder.bat]: Done

File: scripts/build.ps1 (hunk 7 above)
Comment (daxgames) [reply scripts/build.ps1]: No - Removed

**Comment by daxgames** [comment]
@DRSDavidSoft We can't really release 1.4 until https://github.com/microsoft/terminal/issues/16295 gets released.

**Comment by DRSDavidSoft** [comment]
@daxgames Why is the PR closed

**Pull request closed without merging**